gcc/rust/typecheck/rust-tyty-variance-analysis.cc

```diff
@@ -0,0 +1,540 @@
+							Term variance,
+							bool invariant_args)
+{
+  SolutionIndex solution_index = lookup_or_add_type (ref);
+
+  size_t num_lifetimes = subst.get_used_arguments ().get_regions ().size ();
+  size_t num_types = subst.get_substs ().size ();
+
+  for (size_t i = 0; i < num_lifetimes + num_types; ++i)
+    {
+      // TODO: What about variance from other crates?
+      auto variance_i
+	= invariant_args
+	    ? Term::make_transform (variance, V::invariant ())
+	    : Term::make_transform (variance,
+				    Term::make_ref (solution_index + i));
+
+      if (i < num_lifetimes)
+	{
+	  auto region_i = i;
+	  auto &region
+	    = subst.get_substitution_arguments ().get_mut_regions ()[region_i];
+	  add_constraints_from_region (region, variance_i);
+	}
+      else
+	{
+	  auto type_i = i - num_lifetimes;
+	  auto arg = subst.get_arg_at (type_i);
+	  if (arg.has_value ())
+	    {
+	      add_constraints_from_ty (arg.value ().get_tyty (), variance_i);
+	    }
+	}
+    }
+}
+void
+GenericTyVisitorCtx::add_constrints_from_param (ParamType &type, Term variance)
+{
+  auto it
+    = std::find (param_names.begin (), param_names.end (), type.get_name ());
+  rust_assert (it != param_names.end ());
+
+  auto index = first_type + std::distance (param_names.begin (), it);
+
+  add_constraint (index, variance);
+}
+
+Term
+GenericTyVisitorCtx::contra (Term variance)
+{
+  return Term::make_transform (variance, V::contravariant ());
+}
+
+void
+TyVisitorCtx::add_constraints_from_ty (BaseType *ty, Variance variance)
+{
+  Visitor visitor (*this, variance);
+  ty->accept_vis (visitor);
+}
+
+void
+TyVisitorCtx::add_constraints_from_region (const Region &region,
+					   Variance variance)
+{
+  variances.push_back (variance);
+  regions.push_back (region);
+}
+
+void
+TyVisitorCtx::add_constraints_from_generic_args (HirId ref,
+						 SubstitutionRef &subst,
+						 Variance variance,
+						 bool invariant_args)
+{
+  // Handle function
+  auto variances = query_generic_variance (*lookup_type (ref)->as<ADTType> ());
+
+  size_t num_lifetimes = subst.get_used_arguments ().get_regions ().size ();
+  size_t num_types = subst.get_substs ().size ();
+
+  for (size_t i = 0; i < num_lifetimes + num_types; ++i)
+    {
+      // TODO: What about variance from other crates?
+      auto variance_i
+	= invariant_args
+	    ? Variance::transform (variance, Variance::invariant ())
+	    : Variance::transform (variance, variances[i]);
+
+      if (i < num_lifetimes)
+	{
+	  auto region_i = i;
+	  auto &region = subst.get_used_arguments ().get_regions ()[region_i];
+	  add_constraints_from_region (region, variance_i);
+	}
+      else
+	{
+	  auto type_i = i - num_lifetimes;
+	  auto arg = subst.get_arg_at (type_i);
+	  if (arg.has_value ())
+	    {
+	      add_constraints_from_ty (arg.value ().get_tyty (), variance_i);
+	    }
+	}
+    }
+}
+
+Variance
+TyVisitorCtx::contra (Variance variance)
+{
+  return V::transform (variance, V::contravariant ());
```

Comment:
not a big fan of having the `using V` for this line - it would make it clearer to use the full qualified name please